### PR TITLE
gh-98836: Extend PyUnicode_FromFormat()

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
* Support for conversion specifiers o (octal) and X (uppercase hexadecimal).
* Support for length modifiers j (intmax_t) and t (ptrdiff_t).
* Length modifiers are now applied to all integer conversions.
* Support for wchar_t C strings (%ls and %lV).
* Support for variable width and precision (*).
* Support for flag - (left alignment).

* Issue: gh-98836

### CODE REVIEW

File: Doc/whatsnew/3.12.rst [hunk 1]
```diff
@@ -669,12 +669,24 @@ Porting to Python 3.12
...
   :c:func:`PyUnicode_FromFormatV` now sets a :exc:`SystemError`.
   In previous versions it caused all the rest of the format string to be
   copied as-is to the result string, and any extra arguments discarded.
   (Contributed by Serhiy Storchaka in :gh:`95781`.)
 
+An unrecognized format character in :c:func:`PyUnicode_FromFormat` and
+  :c:func:`PyUnicode_FromFormatV` now sets a :exc:`SystemError`.
+  In previous versions it caused all the rest of the format string to be
+  copied as-is to the result string, and any extra arguments discarded.
+  (Contributed by Serhiy Storchaka in :gh:`95781`.)
```
Comment (vstinner) [inline Doc/whatsnew/3.12.rst]: It seems like this paragraph was duplicated by mistake?

File: Lib/test/test_unicode.py [hunk 2]
```diff
@@ -2637,7 +2637,8 @@ def test_from_format(self):
             c_char_p,
             pythonapi, py_object, sizeof,
             c_int, c_long, c_longlong, c_ssize_t,
-            c_uint, c_ulong, c_ulonglong, c_size_t, c_void_p)
+            c_uint, c_ulong, c_ulonglong, c_size_t, c_void_p,
+            sizeof, c_wchar, c_wchar_p)
```
Comment (vstinner) [inline Lib/test/test_unicode.py]: Can you please a comment somewhere to list formats which are not tested and the reason why?

I see at least %j and %t which are not tested. Please mention that _testcapi.test_string_from_format() has a wider coverage of all formats, and test these formats.

File: Lib/test/test_unicode.py [hunk 3]
```diff
@@ -2807,40 +2835,121 @@ def check_format(expected, format, *args):
...
+        check_format('0000000123', b'%010i', c_int(123))
+        check_format('123       ', b'%-10i', c_int(123))
+        check_format('123       ', b'%-010i', c_int(123))
+        check_format('123',        b'%.2i', c_int(123))
+        check_format('0000123',    b'%.7i', c_int(123))
+        check_format('       123', b'%10.2i', c_int(123))
+        check_format('   0000123', b'%10.7i', c_int(123))
+        check_format('0000000123', b'%010.7i', c_int(123))
+        check_format('0000123   ', b'%-10.7i', c_int(123))
+        check_format('0000123   ', b'%-010.7i', c_int(123))
```
Comment (vstinner) [inline Lib/test/test_unicode.py]: Would it be possible to generate these tests? The  code is similar for the 6 groups of tests.

File: Lib/test/test_unicode.py [hunk 4]
```diff
@@ -2807,40 +2835,121 @@ def check_format(expected, format, *args):
...
         # test %V
-        check_format('repr=abc',
-                     b'repr=%V', 'abc', b'xyz')
+        check_format('abc',
+                     b'%V', 'abc', b'xyz')
+        check_format('xyz',
+                     b'%V', None, b'xyz')
+
+        # test %ls
+        check_format('abc', b'%ls', c_wchar_p('abc'))
```
Comment (vstinner) [inline Lib/test/test_unicode.py]: You can please add one non-ASCII character in wchar tests, %ls and %lV? In addition to tests truncating to 2 wchar_t.

File: Lib/test/test_unicode.py [hunk 5]
```diff
@@ -2880,10 +2989,11 @@ def check_format(expected, format, *args):
         # check for crashes
```
Comment (vstinner) [inline Lib/test/test_unicode.py]: this comment is misleading. Python doesn't crash but reject invalid format string with a SyntaxError. Would you mind to rephrase the comment? Like: ``# test invalid format strings``?

File: Doc/c-api/unicode.rst [hunk 6]
```diff
@@ -394,98 +394,149 @@ APIs:
...
-   |                   |                     | ``NULL``).                       |
-   +-------------------+---------------------+----------------------------------+
-   | :attr:`%S`        | PyObject\*          | The result of calling            |
-   |                   |                     | :c:func:`PyObject_Str`.          |
-   +-------------------+---------------------+----------------------------------+
-   | :attr:`%R`        | PyObject\*          | The result of calling            |
-   |                   |                     | :c:func:`PyObject_Repr`.         |
-   +-------------------+---------------------+----------------------------------+
+   ASCII-encoded string.
+
```
Comment (vstinner) [inline Doc/c-api/unicode.rst:398]: Can you please add something like ``[[fill]align][sign][z][#][0][width][grouping_option][.precision][type]``? Example taken from: https://docs.python.org/dev/library/string.html#format-specification-mini-language

The printf format is complex, and it's not obvious in which order each part should be written. I understand that it's something like:

```
%[conversion flags][minimum width][.precision][length modifier][conversion type]
```

Or if you prefer a shorter version:

```
%[conversion][width][.precision][modifier][conversion]
```

Comment (serhiy-storchaka) [reply Doc/c-api/unicode.rst:398]: I copied the following paragraphs from the description of the print-like formatting in Python: https://docs.python.org/3/library/stdtypes.html#printf-style-string-formatting.

I think that it also can benefit from from similar changes, but let leave it to other issue.

File: Doc/whatsnew/3.12.rst (hunk 1 above)
Comment (serhiy-storchaka) [reply Doc/whatsnew/3.12.rst]: True.

File: Lib/test/test_unicode.py (hunk 2 above)
Comment (serhiy-storchaka) [reply Lib/test/test_unicode.py]: Done.

File: Lib/test/test_unicode.py (hunk 3 above)
Comment (serhiy-storchaka) [reply Lib/test/test_unicode.py]: It is not so similar. It is different for signed and unsigned types, but I'll try to generalize it.

File: Lib/test/test_unicode.py (hunk 4 above)
Comment (serhiy-storchaka) [reply Lib/test/test_unicode.py]: What do you mean? Isn't the following line enough?

File: Lib/test/test_unicode.py (hunk 5 above)
Comment (serhiy-storchaka) [reply Lib/test/test_unicode.py]: Done. Restored the part of the older comment.

File: Lib/test/test_unicode.py (hunk 3 above)
Comment (vstinner) [reply Lib/test/test_unicode.py]: I just suggested to generate these tests. It's ok to leave them as they are if it's too complicated to generate them.

File: Lib/test/test_unicode.py [hunk 7]
```diff
@@ -2632,12 +2632,17 @@ class CAPITest(unittest.TestCase):
 
     # Test PyUnicode_FromFormat()
     def test_from_format(self):
+        # Length modifiers "j" and "t" are not tested here because ctypes does
+        # not expose types for intmax_t and ptrdiff_t.
+        # _testcapi.test_string_from_format() has a wider coverage of all
+        # formats.
```
Comment (vstinner) [inline Lib/test/test_unicode.py]: Thank you, this comment is useful to me at least :-)

File: Lib/test/test_unicode.py (hunk 4 above)
Comment (vstinner) [reply Lib/test/test_unicode.py]: I'm thinking at adding a test with non-ASCII characters which fit into UCS-2 (16-bit wchar_t) characters. Something like:

```
check_format('a\xe9\u20ac', b'%ls', c_wchar_p('a\xe9\u20ac'))
```

Some bugs are only triggered depending on the maximum code point, and wchar_t code can take different code path depending if there is a surrogate character or not.

**Review by vstinner — APPROVED** [review]
LGTM.

**Comment by vstinner** [comment]
Feel free to ignore my suggestions, the PR LGTM.

File: Lib/test/test_unicode.py (hunk 4 above)
Comment (serhiy-storchaka) [reply Lib/test/test_unicode.py]: Done.

**Review by vstinner — APPROVED** [review]
LGTM.

**Pull request merged**